### PR TITLE
fix(chai-retry-plugin): total test time and tests

### PR DESCRIPTION
**Pull request by vladyslav-baliuk**
Improvements for `chai-retry-plugin`:

- total test time should be adjusted according to all timeouts and delays
- cover more `chai`'s build-in asserters in tests

### CODE REVIEW

**Comment by yurii-ve** [comment]
Missing PR description

File: packages/testing/src/chai-retry-plugin/helpers.ts [hunk 1]
```diff
@@ -1,9 +1,22 @@
 import Chai from 'chai';
-import { timeout as timeoutPromise, sleep } from 'promise-assist';
+import { timeout as timeoutPromise, sleep as sleepPromise } from 'promise-assist';
```
Comment (yurii-ve) [inline packages/testing/src/chai-retry-plugin/helpers.ts:2]: it's a matter of opinion of course, but in such cases I'd not destructure, but instead did this:
```
import promiseHelpers from 'promise-assist';
...
promiseHelpers.sleep();
```

Also `sleepPromise` is a bad name, because it's function that returns promise - not promise.

File: packages/testing/src/chai-retry-plugin/helpers.ts [hunk 2]
```diff
@@ -1,9 +1,22 @@
...
+import { adjustTestTime, mochaCtx } from '../mocha-ctx';
 
 import { chaiMethodsThatHandleFunction } from './constants';
 import type { RetryAndAssertArguments } from './types';
 
+export const addTimeoutSafetyMargin = (ms: number) => mochaCtx() && adjustTestTime(ms);
```
Comment (yurii-ve) [inline packages/testing/src/chai-retry-plugin/helpers.ts:8]: It's "half-clear" what this is for... at least for me.
what do you think about leaving a short comment which explains that?

File: packages/testing/src/chai-retry-plugin/helpers.ts [hunk 3]
```diff
@@ -1,9 +1,22 @@
...
+
+function sleep(ms: number): Promise<void> {
```
Comment (yurii-ve) [inline packages/testing/src/chai-retry-plugin/helpers.ts:10]: maybe `sleepWithSafetyMarginTimeout` or something in that nature?

File: packages/testing/src/test/chai-retry-plugin.unit.ts [hunk 4]
```diff
@@ -217,6 +217,30 @@ describe('chai-retry-plugin', () => {
...
+                    return;
+                };
+            }, 900);
+
+            await expect(() => new EchoService(), 'should respond')
+                .retry({ timeout: 1000 })
+                .to.respondTo('echo');
+        });
+
+        it('.keys()', async () => {
```
Comment (yurii-ve) [inline packages/testing/src/test/chai-retry-plugin.unit.ts:237]: what exactly is being tested here?

File: packages/testing/src/test/chai-retry-plugin.unit.ts [hunk 5]
```diff
@@ -217,6 +217,30 @@ describe('chai-retry-plugin', () => {
...
+            const arrayWithKeys = [{}, {}, {}, { a: 1, b: 2 }, { a: 1, b: 2, c: 3 }, { a: 1, b: 2, c: 3, d: 4 }];
+            const { resultFunction, getCallCount } = withCallCount((count) => arrayWithKeys[count]);
```
Comment (yurii-ve) [inline packages/testing/src/test/chai-retry-plugin.unit.ts:239]: there are many tests in this suite with code like this: `if (count === 3) return {a: 1, b: 2, c: 3}`. Why don't do the same here? I mean this array stuff is just confusing... without any benefit in my opinion.

File: packages/testing/src/test/chai-retry-plugin.unit.ts [hunk 6]
```diff
@@ -217,6 +217,30 @@ describe('chai-retry-plugin', () => {
...
             expect(getCallCount()).to.equal(2);
         });
+
+        it('.respondTo()', async () => {
+            class EchoService {
+                echo: (() => void) | undefined;
+            }
+
+            setTimeout(() => {
+                EchoService.prototype.echo = () => {
```
Comment (yurii-ve) [inline packages/testing/src/test/chai-retry-plugin.unit.ts:227]: do we actually need this class/prototype stuff? Will `const o = {}; ... o.test = () => null;` work?

**Comment by yurii-ve** [comment]
there are tests added in this pR but none of them is actually testing that timeout safety margin works.

File: packages/testing/src/chai-retry-plugin/helpers.ts (hunk 2 above)
Comment (vladyslav-baliuk) [reply packages/testing/src/chai-retry-plugin/helpers.ts:10]: @yurii-ve fixed, didn't add it cause imported `mochaCtx` and `adjustTestTime` already had comments with explanations available with InteliSense

File: packages/testing/src/chai-retry-plugin/helpers.ts (hunk 1 above)
Comment (vladyslav-baliuk) [reply packages/testing/src/chai-retry-plugin/helpers.ts]: @yurii-ve nice idea, fixed

File: packages/testing/src/chai-retry-plugin/helpers.ts (hunk 3 above)
Comment (vladyslav-baliuk) [reply packages/testing/src/chai-retry-plugin/helpers.ts]: @yurii-ve fixed

Also renamed `timeout` to `timeoutWithSafetyMargin`

File: packages/testing/src/test/chai-retry-plugin.unit.ts (hunk 4 above)
Comment (vladyslav-baliuk) [reply packages/testing/src/test/chai-retry-plugin.unit.ts:233]: @yurii-ve added this test to cover more `chai`'s built-in asserters, `.keys()` in this case

You are right, `testArray` sounds better) Also removed redundant empty objects

File: packages/testing/src/test/chai-retry-plugin.unit.ts (hunk 5 above)
Comment (vladyslav-baliuk) [reply packages/testing/src/test/chai-retry-plugin.unit.ts]: @yurii-ve added it here just to also test empty obj and partially correct result, you think there is no sense to verify such cases?

File: packages/testing/src/test/chai-retry-plugin.unit.ts (hunk 6 above)
Comment (vladyslav-baliuk) [reply packages/testing/src/test/chai-retry-plugin.unit.ts]: @yurii-ve yeah, could be simplified, fixed

File: packages/testing/src/chai-retry-plugin/helpers.ts (hunk 3 above)
Comment (vladyslav-baliuk) [reply packages/testing/src/chai-retry-plugin/helpers.ts]: > Missing PR description

fixed

**Comment by vladyslav-baliuk** [comment]
> Missing PR description

fixed

**Comment by vladyslav-baliuk** [comment]
> there are tests added in this PR but none of them is actually testing that timeout safety margin works.

I guess there is no need to add such test in this exact test-suite, cause I'm using functions `mochaCtx` and `adjustTestTime` that are already created and tested in `mocha.unit.ts`. Or that would be valuable anyway?

File: packages/testing/src/test/chai-retry-plugin.unit.ts (hunk 5 above)
Comment (yurii-ve) [reply packages/testing/src/test/chai-retry-plugin.unit.ts]: in my opinion with such cases test that keys assertion works correctly and they are simply redundant if purpose of the test is to verify that retry plugin works correctly with keys assertion

**Comment by yurii-ve** [comment]
> > there are tests added in this PR but none of them is actually testing that timeout safety margin works.
> 
> I guess there is no need to add such test in this exact test-suite, cause I'm using functions `mochaCtx` and `adjustTestTime` that are already created and tested in `mocha.unit.ts`. Or that would be valuable anyway?

This is a topic for discussion of course, but my reasoning is that there is code that adds this timeout safety margin in the retry-plugin implementation and would be nice to have a failing test if somebody removed this code.

File: packages/testing/src/test/chai-retry-plugin.unit.ts (hunk 5 above)
Comment (vladyslav-baliuk) [reply packages/testing/src/test/chai-retry-plugin.unit.ts]: @yurii-ve fixed

**Comment by vladyslav-baliuk** [comment]
> > > there are tests added in this PR but none of them is actually testing that timeout safety margin works.
> > 
> > 
> > I guess there is no need to add such test in this exact test-suite, cause I'm using functions `mochaCtx` and `adjustTestTime` that are already created and tested in `mocha.unit.ts`. Or that would be valuable anyway?
> 
> This is a topic for discussion of course, but my reasoning is that there is code that adds this timeout safety margin in the retry-plugin implementation and would be nice to have a failing test if somebody removed this code.

added test for that 👍

**Review by yurii-ve — APPROVED** [review]

File: packages/testing/src/test/chai-retry-plugin.unit.ts [hunk 7]
```diff
@@ -58,6 +58,18 @@ describe('chai-retry-plugin', () => {
                 expect(getCallCount()).to.be.within(9, 10);
             }
         });
+
+        it(`should work when mocha's timeout less then time needed for test`, async function () {
```
[the remark this answers is not in the document]
Comment (vladyslav-baliuk) [reply packages/testing/src/test/chai-retry-plugin.unit.ts]: @yurii-ve yep :)
